File: lib/core/models/candle.dart
/// Represents a single OHLCV candlestick.
///
/// This is the canonical candle model used throughout the Flutter app.
/// It can be constructed from:
///   - Binance REST API kline arrays ([fromBinanceKline])
///   - JSON maps ([fromJson]) for cache deserialization
///
/// And converted to:
///   - JSON maps ([toJson]) for cache serialization
///   - Rust-engine-compatible JSON ([toRustJson]) for the FFI bridge
///
/// ## Example – Fetch BTC/USDT 1h data for the last 30 days
/// ```dart
/// final client = BinanceApiClient();
/// final candles = await client.downloadHistory(
///   symbol: 'BTCUSDT',
///   interval: '1h',
///   days: 30,
/// );
/// // Convert for Rust engine
/// final rustJson = candles.map((c) => c.toRustJson()).toList();
/// ```
class CandleData {
  /// Unix timestamp in milliseconds (UTC, candle open time).
  final int timestamp;

  /// Opening price.
  final double open;

  /// Highest price during the period.
  final double high;

  /// Lowest price during the period.
  final double low;

  /// Closing price.
  final double close;

  /// Base-asset trading volume during the period.
  final double volume;

  const CandleData({
    required this.timestamp,
    required this.open,
    required this.high,
    required this.low,
    required this.close,
    required this.volume,
  });

  // ─── Factory constructors ─────────────────────────────────────────────

  /// Parse a Binance kline array.
  ///
  /// Binance REST `/api/v3/klines` returns each candle as a JSON array:
  /// ```
  /// [
  ///   1499040000000,      // 0  Open time (ms)
  ///   "0.01634000",       // 1  Open
  ///   "0.80000000",       // 2  High
  ///   "0.01575800",       // 3  Low
  ///   "0.01577100",       // 4  Close
  ///   "148976.11427815",  // 5  Volume
  ///   1499644799999,      // 6  Close time (ms)
  ///   "2434.19055334",    // 7  Quote asset volume
  ///   308,                // 8  Number of trades
  ///   "1756.87402397",    // 9  Taker buy base volume
  ///   "28.46694368",      // 10 Taker buy quote volume
  ///   "17928899.62484339" // 11 Ignore
  /// ]
  /// ```
  factory CandleData.fromBinanceKline(List<dynamic> kline) {
    return CandleData(
      timestamp: kline[0] as int,
      open: double.parse(kline[1] as String),
      high: double.parse(kline[2] as String),
      low: double.parse(kline[3] as String),
      close: double.parse(kline[4] as String),
      volume: double.parse(kline[5] as String),
    );
  }

  /// Alias kept for backward compatibility.
  factory CandleData.fromBinanceJson(List<dynamic> json) =
      CandleData.fromBinanceKline;

  /// Deserialize from a JSON map (e.g. from cache).
  factory CandleData.fromJson(Map<String, dynamic> json) {
    return CandleData(
      timestamp: json['timestamp'] as int,
      open: (json['open'] as num).toDouble(),
      high: (json['high'] as num).toDouble(),
      low: (json['low'] as num).toDouble(),
      close: (json['close'] as num).toDouble(),
      volume: (json['volume'] as num).toDouble(),
    );
  }

  // ─── Serialization ────────────────────────────────────────────────────

  /// Serialize to a JSON map (for caching / generic use).
  Map<String, dynamic> toJson() => {
        'timestamp': timestamp,
        'open': open,
        'high': high,
        'low': low,
        'close': close,
        'volume': volume,
      };

  /// Serialize to the exact JSON format expected by the Rust trading engine.
  ///
  /// The Rust `Candle` struct uses the same field names, so this is
  /// identical to [toJson] but exists as a named method for clarity.
  Map<String, dynamic> toRustJson() => toJson();

  // ─── Derived properties ───────────────────────────────────────────────

  /// The candle open time as a [DateTime] (UTC).
  DateTime get dateTime =>
      DateTime.fromMillisecondsSinceEpoch(timestamp, isUtc: true);

  /// Whether this candle is bullish (close ≥ open).
  bool get isBullish => close >= open;

  /// Whether this candle is bearish (close < open).
  bool get isBearish => close < open;

  /// Absolute body size |close − open|.
  double get bodySize => (close - open).abs();

  /// Full range (high − low).
  double get range => high - low;

  /// Mid-price (high + low) / 2.
  double get midpoint => (high + low) / 2.0;

  // ─── Equality & display ───────────────────────────────────────────────

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is CandleData &&
          timestamp == other.timestamp &&
          open == other.open &&
          high == other.high &&
          low == other.low &&
          close == other.close &&
          volume == other.volume;

  @override
  int get hashCode => Object.hash(timestamp, open, high, low, close, volume);

  @override
  String toString() =>
      'CandleData(ts=$timestamp, O=$open, H=$high, L=$low, C=$close, V=$volume)';
}

File: rust/trading_engine/src/models/candle.rs
use serde::{Deserialize, Serialize};

/// Represents a single OHLCV candlestick.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Candle {
    /// Unix timestamp in milliseconds.
    pub timestamp: i64,
    /// Opening price.
    pub open: f64,
    /// Highest price during the period.
    pub high: f64,
    /// Lowest price during the period.
    pub low: f64,
    /// Closing price.
    pub close: f64,
    /// Trading volume.
    pub volume: f64,
}

impl Candle {
    /// Create a new Candle.
    pub fn new(timestamp: i64, open: f64, high: f64, low: f64, close: f64, volume: f64) -> Self {
        Self {
            timestamp,
            open,
            high,
            low,
            close,
            volume,
        }
    }

    /// Returns whether this candle is bullish (close >= open).
    pub fn is_bullish(&self) -> bool {
        self.close >= self.open
    }

    /// Returns the body size (absolute difference between open and close).
    pub fn body_size(&self) -> f64 {
        (self.close - self.open).abs()
    }

    /// Returns the full range (high - low).
    pub fn range(&self) -> f64 {
        self.high - self.low
    }

    /// Returns the midpoint price ((high + low) / 2).
    pub fn midpoint(&self) -> f64 {
        (self.high + self.low) / 2.0
    }
}

/// Supported timeframes for candle data.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum Timeframe {
    M1,
    M5,
    M15,
    M30,
    H1,
    H4,
    D1,
    W1,
}

impl Timeframe {
    /// Returns the timeframe string used by exchanges (e.g., "1m", "1h").
    pub fn as_str(&self) -> &'static str {
        match self {
            Timeframe::M1 => "1m",
            Timeframe::M5 => "5m",
            Timeframe::M15 => "15m",
            Timeframe::M30 => "30m",
            Timeframe::H1 => "1h",
            Timeframe::H4 => "4h",
            Timeframe::D1 => "1d",
            Timeframe::W1 => "1w",
        }
    }

    /// Returns the duration of this timeframe in milliseconds.
    pub fn duration_ms(&self) -> i64 {
        match self {
            Timeframe::M1 => 60_000,
            Timeframe::M5 => 300_000,
            Timeframe::M15 => 900_000,
            Timeframe::M30 => 1_800_000,
            Timeframe::H1 => 3_600_000,
            Timeframe::H4 => 14_400_000,
            Timeframe::D1 => 86_400_000,
            Timeframe::W1 => 604_800_000,
        }
    }
}

/// Container for OHLCV data: a collection of candles for a specific symbol and timeframe.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OhlcvData {
    /// Trading symbol (e.g., "BTCUSDT").
    pub symbol: String,
    /// Timeframe of the candle data.
    pub timeframe: Timeframe,
    /// Ordered list of candles (oldest first).
    pub candles: Vec<Candle>,
}

impl OhlcvData {
    /// Create a new empty OhlcvData container.
    pub fn new(symbol: String, timeframe: Timeframe) -> Self {
        Self {
            symbol,
            timeframe,
            candles: Vec::new(),
        }
    }

    /// Append a candle to the data.
    pub fn push(&mut self, candle: Candle) {
        self.candles.push(candle);
    }

    /// Returns the number of candles.
    pub fn len(&self) -> usize {
        self.candles.len()
    }

    /// Returns true if there are no candles.
    pub fn is_empty(&self) -> bool {
        self.candles.is_empty()
    }

    /// Returns a slice of the last `n` candles.
    pub fn last_n(&self, n: usize) -> &[Candle] {
        let start = self.candles.len().saturating_sub(n);
        &self.candles[start..]
    }

    /// Returns the latest candle if available.
    pub fn latest(&self) -> Option<&Candle> {
        self.candles.last()
    }

    /// Returns a slice of close prices.
    pub fn closes(&self) -> Vec<f64> {
        self.candles.iter().map(|c| c.close).collect()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_candle_new() {
        let candle = Candle::new(1000, 100.0, 110.0, 95.0, 105.0, 500.0);
        assert_eq!(candle.timestamp, 1000);
        assert_eq!(candle.open, 100.0);
        assert_eq!(candle.high, 110.0);
        assert_eq!(candle.low, 95.0);
        assert_eq!(candle.close, 105.0);
        assert_eq!(candle.volume, 500.0);
    }

    #[test]
    fn test_candle_bullish() {
        let bullish = Candle::new(1000, 100.0, 110.0, 95.0, 105.0, 500.0);
        assert!(bullish.is_bullish());

        let bearish = Candle::new(1000, 105.0, 110.0, 95.0, 100.0, 500.0);
        assert!(!bearish.is_bullish());

        let doji = Candle::new(1000, 100.0, 110.0, 95.0, 100.0, 500.0);
        assert!(doji.is_bullish()); // close == open is considered bullish
    }

    #[test]
    fn test_candle_body_size() {
        let candle = Candle::new(1000, 100.0, 110.0, 95.0, 105.0, 500.0);
        assert!((candle.body_size() - 5.0).abs() < f64::EPSILON);
    }

    #[test]
    fn test_candle_range() {
        let candle = Candle::new(1000, 100.0, 110.0, 95.0, 105.0, 500.0);
        assert!((candle.range() - 15.0).abs() < f64::EPSILON);
    }

    #[test]
    fn test_timeframe_as_str() {
        assert_eq!(Timeframe::M1.as_str(), "1m");
        assert_eq!(Timeframe::H1.as_str(), "1h");
        assert_eq!(Timeframe::D1.as_str(), "1d");
    }

    #[test]
    fn test_ohlcv_data() {
        let mut data = OhlcvData::new("BTCUSDT".to_string(), Timeframe::H1);
        assert!(data.is_empty());
        assert_eq!(data.len(), 0);

        data.push(Candle::new(1000, 100.0, 110.0, 95.0, 105.0, 500.0));
        data.push(Candle::new(2000, 105.0, 115.0, 100.0, 110.0, 600.0));

        assert_eq!(data.len(), 2);
        assert!(!data.is_empty());
        assert_eq!(data.latest().unwrap().close, 110.0);
        assert_eq!(data.closes(), vec![105.0, 110.0]);
        assert_eq!(data.last_n(1).len(), 1);
        assert_eq!(data.last_n(1)[0].close, 110.0);
    }

    #[test]
    fn test_candle_serialization() {
        let candle = Candle::new(1000, 100.0, 110.0, 95.0, 105.0, 500.0);
        let json = serde_json::to_string(&candle).unwrap();
        let deserialized: Candle = serde_json::from_str(&json).unwrap();
        assert_eq!(candle, deserialized);
    }
}
